doc/developer/logging.rst

```diff
@@ -239,3 +242,180 @@ Examples:
 * some field that is absolutely needed is :code:`NULL`
 * any other kind of data structure corruption that will cause the daemon
   to crash sooner or later, one way or another
+
+Thread-local buffering
+----------------------
+
+The core logging code in :file:`lib/zlog.c` allows setting up per-thread log
+message buffers in order to improve logging performance.  The following rules
+apply for this buffering:
+
+* Only messages of priority *DEBUG* or *INFO* are buffered.
+* Any higher-priority message causes the thread's entire buffer to be flushed,
+  thus message ordering is preserved on a per-thread level.
+* There is no guarantee on ordering between different threads;  in most cases
+  this is arbitrary to begin with since the threads essentially race each
+  other in printing log messages.  If an order is established with some
+  synchronization primitive, add calls to :c:func:`zlog_tls_buffer_flush()`.
+* The buffers are only ever accessed by the thread they are created by.  This
+  means no locking is necessary.
+
+Both the main/default thread and additional threads created by
+:c:func:`frr_pthread_new()` with the default :c:func:`frr_run()` handler will
+initialize thread-local buffering and call :c:func:`zlog_tls_buffer_flush()`
+when idle.
+
+If some piece of code runs for an extended period, it may be useful to insert
+calls to :c:func:`zlog_tls_buffer_flush()` in appropriate places:
+
+.. c:function:: void zlog_tls_buffer_flush(void)
+
+   Write out any pending log messages that the calling thread may have in its
+   buffer.  This function is safe to call regardless of the per-thread log
+   buffer being set up / in use or not.
```

Comment:
I've thought about it and doing an explicit timeout is somewhat complicated... however, there is a `zlog_tls_buffer_flush()` in the `thread_master` code so it's not possible for a thread to go idle with log messages still in the buffer.

The major case where this would matter is an infinite loop somewhere after printing some debug messages.  In this case, buffered messages can be read from `/var/tmp` like in the SEGV case.